provider/datagen/src/transform/cldr/units/mod.rs

```diff
@@ -2,34 +2,137 @@
-use zerovec::ZeroMap;
+use icu_unitsconversion::provider::{
+    ConstantType, ConstantValue, UnitsConstantsV1, UnitsConstantsV1Marker,
+};
+use zerovec::{ZeroMap, ZeroVec};
 
 impl DataProvider<UnitsConstantsV1Marker> for crate::DatagenProvider {
     fn load(&self, _req: DataRequest) -> Result<DataResponse<UnitsConstantsV1Marker>, DataError> {
         self.check_req::<UnitsConstantsV1Marker>(_req)?;
 
-        let _units_data: &cldr_serde::units::units_constants::Resource = self
+        let units_data: &cldr_serde::units::units_constants::Resource = self
             .cldr()?
             .core()
             .read_and_parse("supplemental/units.json")?;
-        let mut constants_map = BTreeMap::<&str, &str>::new();
+        let constants = &units_data.supplemental.unit_constants.constants;
+
+        let mut constants_map_in_str_form =
+            BTreeMap::<&str, (Vec<String>, Vec<String>, ConstantType)>::new();
+        for (cons_name, cons_value) in constants {
+            let value = remove_whitespace(&cons_value.value);
+            let (num, den) = split_constant_string(&value)?;
 
-        let constants = &_units_data.supplemental.unit_constants.constants;
-        for (key, constant) in constants {
-            constants_map.insert(key, &constant.value);
+            let constant_type = match cons_value.status.as_deref() {
+                Some("approximate") => ConstantType::Approximate,
+                _ => ConstantType::Actual,
+            };
+
+            constants_map_in_str_form.insert(cons_name, (num, den, constant_type));
         }
 
-        let result = UnitsConstantsV1 {
-            constants_map: ZeroMap::from_iter(constants_map),
-        };
+        // This loop iterates over the constants, replacing any string values with their corresponding constant values.
+        let mut num_of_const_with_text;
+        loop {
+            num_of_const_with_text = 0;
+            let mut constants_with_constants_map_replaceable =
+                BTreeMap::<&str, (Vec<String>, Vec<String>, ConstantType)>::new();
+            for (cons_name, (num, den, constant_type)) in constants_map_in_str_form.iter() {
+                let mut temp_num = num.clone();
+                let mut temp_den = den.clone();
+                let mut temp_constant_type = *constant_type;
+
+                for i in 0..temp_num.len() {
+                    if !contains_alphabetic_chars(temp_num[i].as_str())
+                        || is_scientific_number(temp_num[i].as_str())
+                    {
+                        continue;
+                    }
+
+                    num_of_const_with_text += 1;
+                    if let Some((rnum, rden, rconstant_type)) =
+                        constants_map_in_str_form.get(temp_num[i].as_str())
+                    {
+                        temp_num.remove(i);
+                        // append the elements in rnum to num and rden to den
+                        temp_num.append(&mut rnum.clone());
```

Comment:
this clones `rnum` into a temporary vector, and then clears that and moves the elements into `temp_num` (which is why you need the `&mut`). This is quite a roundabout way of doing this, consider
```suggestion
                        temp_num.extend(rnum.iter().cloned());
```

Reply:
`temp_num.extend(rnum.clone().into_iter())`

Reply:
That has the same problem of allocating an intermediate `Vec<String>`